merlin/core/compat.py

```diff
@@ -68,6 +130,69 @@ def _get_gpu_count():
 
 try:
     import tensorflow
+
+    def configure_tensorflow(memory_allocation=None, device=None):
+        """Utility to help configure tensorflow to not use 100% of gpu memory as buffer"""
+        tf = tensorflow
+        total_gpu_mem_mb = device_mem_size(kind="total", cpu=(not HAS_GPU)) / (1024**2)
+
+        if memory_allocation is None:
+            memory_allocation = os.environ.get("TF_MEMORY_ALLOCATION", 0.5)
+
+        if float(memory_allocation) < 1:
+            memory_allocation = total_gpu_mem_mb * float(memory_allocation)
+        memory_allocation = int(memory_allocation)
+        assert memory_allocation < total_gpu_mem_mb
+
+        # TODO: what will this look like in any sort
+        # of distributed set up?
+        if device is None:
+            device = int(os.environ.get("TF_VISIBLE_DEVICE", 0))
+        tf_devices = tf.config.list_physical_devices("GPU")
+        if HAS_GPU and len(tf_devices) == 0:
+            raise ImportError("TensorFlow is not configured for GPU")
+        if HAS_GPU:
+            try:
+                tf.config.set_logical_device_configuration(
+                    tf_devices[device],
+                    [tf.config.LogicalDeviceConfiguration(memory_limit=memory_allocation)],
+                )
+            except RuntimeError:
+                warnings.warn(
+                    "TensorFlow runtime already initialized, may not be enough memory for cudf"
+                )
+            try:
+                tf.config.experimental.set_virtual_device_configuration(
+                    tf_devices[device],
+                    [
+                        tf.config.experimental.VirtualDeviceConfiguration(
+                            memory_limit=memory_allocation
+                        )
+                    ],
+                )
+            except RuntimeError as e:
+                # Virtual devices must be set before GPUs have been initialized
+                warnings.warn(str(e))
+
+        # versions using TF earlier than 2.3.0 need to use extension
+        # library for dlpack support to avoid memory leak issue
+        __TF_DLPACK_STABLE_VERSION = "2.3.0"
+        if version.parse(tf.__version__) < version.parse(__TF_DLPACK_STABLE_VERSION):
+            try:
+                from tfdlpack import from_dlpack
+            except ModuleNotFoundError as e:
+                message = (
+                    "If using TensorFlow < 2.3.0, you must install tfdlpack-gpu extension library"
+                )
+                raise ModuleNotFoundError(message) from e
+
+        else:
+            from tensorflow.experimental.dlpack import from_dlpack
+
+        return from_dlpack
+
+    configure_tensorflow()
```

Comment:
would it be clearer to call this functoin in the libraries that need it instead of in the global scope?

Reply:
this tensorflow import wasn't added in this PR, however it feels like a relevant place to discuss this. We seem to be following a path with this compat module to having a side effect of importing every framework module to use any core feature. For example to import the Dataset or Schema, we now have a side effect of importing and configuring tensorflow (if it's installed). Tensorflow is a good example of a package that is not  lightweight from an import perspective. I wonder how we might refacor this so that we can optionally import tensorflow only when required.